Enabling "service" watch handler to accept a slice of tags

dcarbone opened the pull request:
Hey there,

This PR enables a user to watch  service with multiple tags, utilizing the new multi-tag feature of the health service endpoint.

Valid input could be one of:

```golang
var params map[string]interface{}

// existing functionality is unimpeded
params = map[string]interface{}{"type":"service", "service":"foo", "tag":"bar", "passingonly":true}

// added new "tags" param
params = map[string]interface{}{"type":"service", "service":"foo", "tags":["buzz","biff"], "passingonly":true}

// can combine the two, if you want
params = map[string]interface{}{"type":"service", "service":"foo", "tag": "bar", "tags":["buzz","biff"], "passingonly":true}
// this results in "tags": {"buzz","biff","bar"}
```

I'm fairly certain the implementation could be improved, so comments welcome.

Tests pass tho, so prod ready, yes?

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=5347)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=5347)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=5347).</sub>

dcarbone commented:
I doubt my changes effect cluster leader election, but of course if it did somehow I will happily update it.  I will look later this afternoon.

mkeeler commented on watch/funcs.go:
```diff
@@ -155,7 +170,7 @@ func serviceWatch(params map[string]interface{}) (WatcherFunc, error) {
 		opts := makeQueryOptionsWithContext(p, stale)
 		defer p.cancelFunc()
-		nodes, meta, err := health.Service(service, tag, passingOnly, &opts)
+		nodes, meta, err := health.ServiceMultipleTags(service, tags, passingOnly, &opts)
```
One issue here is that this transparently breaks compatibility with older versions of Consul.

There are a few ways to address this but I think one easy way is to use the `health.Service` method when `tag != ""` and tags originally was nil. Basically if the watcher is using the tags array it would implicitly opt-in to using the newer incompatible APIs.

dcarbone replied:
will do

dcarbone replied:
@mkeeler sorry I misunderstood your objection at first.

I do not believe this actually is a BC break given that the current implementation of the `Health` service itself does this:
```golang
// Service is used to query health information along with service info
// for a given service. It can optionally do server-side filtering on a tag
// or nodes with passing health checks only.
func (h *Health) Service(service, tag string, passingOnly bool, q *QueryOptions) ([]*ServiceEntry, *QueryMeta, error) {
	var tags []string
	if tag != "" {
		tags = []string{tag}
	}
	return h.service(service, tags, passingOnly, q, false)
}
```

The underlying `service` method expects `tags []string`, not `tag string`.  Additionally, given that the end result is `?tag=tag1&tag=tag2&tag=tag3`, I do not think its possible for this to be an actual BC break.

Of course, please correct me if I am wrong...

dcarbone replied:
@mkeeler: Could you please elaborate the BC component to my PR?  I don't believe it actually does.  I have not had a chance to spin up a old 0.6.4 agent or something to test against, but I can spend some time later today doing so and provide the logs if it will help move this PR along.

mkeeler replied:
@dcarbone After digging a little deeper, I think you are right. The API and HTTP layer of Consul just use multiple `tag` params to specify a multi-tag filter. Its when this gets translated into RPCs by the HTTP server where things could go poorly but that is already handled and not something you need to worry about with this PR.

mkeeler commented on watch/funcs_test.go:
```diff
@@ -357,6 +357,102 @@ func TestServiceWatch(t *testing.T) {
 
+func TestServiceMultipleTagsWatch(t *testing.T) {
+	t.Parallel()
+	a := agent.NewTestAgent(t.Name(), ``)
```
```suggestion
	a := agent.NewTestAgent(t, t.Name(), ``)
```

I just merged a change this morning that passes the `t` value into the TestAgent. This is the cause of the current test failure.

dcarbone replied:
updated master merged and pushed :)

dcarbone commented:
@mkeeler After looking at the code and other implementations, I'd like a decision on some implementation options:

#### 1. Use CSV
```
consul watch -type=service -service=redis -tag=mytag -tags="mytag2,mytag3"
```
#### 2. Use the multiple definition option as used by the acl policy id flag:
```
consul watch -type=service -service=redis -tag=mytag -tag=mytag2 -tag=mytag3
```
#### 3. Use two separate tags, one allowing for multi definition and one not
```
consul watch -type=service -service=redis -tag=mytag -tags=mytag2 -tags=mytag3
```

I personally think option 2 is the most idiomatic choice for this.  It would maintain existing functionality while adding the multi-value flag paradigm set by other commands, such as the ACL token command.

I will, of course, implement whichever.

dcarbone commented:
@mkeeler do you have an opinion on which to implement?  if not I will go ahead with option 2.

mkeeler commented:
@dcarbone Sorry for missing that 5 days ago now. I think 2 seems right. It maintains compatibility with anything using it today (they can still do a -tag abc) while allowing us to collect the full set of tags within a slice.

dcarbone commented:
@mkeeler: apologies for the delay, work got hectic.

After some thinking, my initial implementation was short sighted.  I think it would ultimately be confusing to have the json / hcl have a "tags" field and the cli to just have a "tag" option.  To that end, I have updated the func itself to accept "tag" as either a string, []string, or []interface{}.  I have also updated the docs and the cli command itself.

Let me know if you have any questions / suggestions!

mkeeler commented on website/source/docs/agent/watches.html.md:
```diff
@@ -266,26 +266,48 @@ An example of the output of this command:
-The "tag" parameter will filter by tag, and "passingonly" is
-a boolean that will filter to only the instances passing all
-health checks.
+and optionally takes the parameters "tag", "tags", and 
```
Are the docs here still accurate after your latest updates?

dcarbone replied:
Ah!  my bad, i updated the top statement but forgot about that one.  Apologies!

mkeeler replied:
No need to apologize. Thanks for all your work on this.

dcarbone replied:
@mkeeler fixed.

mkeeler commented:
Thanks again @dcarbone. There were a handful of merge conflicts I introduce when I moved the watch package recently so I branched from your PR to make those fixes and this has now been merged.
